karmada-io/karmada
# Introduce a mechanism to scheduler to actively trigger rescheduling

**What type of PR is this?**

/kind api-change
/kind feature



**What this PR does / why we need it**:

Introduce a mechanism to scheduler to actively trigger rescheduling

**Which issue(s) this PR fixes**:

Fixes part of #4840

**Special notes for your reviewer**:

How to **test** it：

1. create a deployment and policy like follows:

<details>
<summary>resource.yaml</summary>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: demo-test-1
  labels:
    app: test
spec:
  replicas: 3
  selector:
    matchLabels:
      app: demo-test-1
  template:
    metadata:
      labels:
        app: demo-test-1
    spec:
      terminationGracePeriodSeconds: 0
      containers:
        - image: nginx
          name: demo-test-1
          resources:
            limits:
              cpu: 10m
              memory: 10Mi
---
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: default-pp
spec:
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
    replicaScheduling:
      replicaDivisionPreference: Weighted
      replicaSchedulingType: Divided
      weightPreference:
        dynamicWeight: AvailableReplicas
  resourceSelectors:
    - apiVersion: apps/v1
      kind: Deployment
      name: demo-test-1
      namespace: default
```
</details>

2. when first schedule finished, update schedule result in bingding manually, like:

```shell
ka patch rb demo-test-1-deployment --type='json' -p '[{"op": "replace", "path": "/spec/clusters", "value": [{"name": "member2", "replicas": 3}]}]'
```

3. add filed `spec.rescheduleTriggeredAt` binding manually, like:

```shell
ka patch rb demo-test-1-deployment --type='json' -p '[{"op": "replace", "path": "/spec/rescheduleTriggeredAt", "value": "2024-04-18T11:38:15Z"}]'
```

> notes that the timestamp value can set to one second later than current `status.lastScheduledTime`.

4. check whether reschedule executed by observing whether `spec.cluster` and `status.lastScheduledTime` updated.

**Does this PR introduce a user-facing change?**:

```release-note
Introduce a mechanism to scheduler to actively trigger rescheduling
```
